File: 0874. Walking Robot Simulation/0874_myans.cpp
class Solution
{
public:
    int robotSim(vector<int> &commands, vector<vector<int>> &obstacles)
    {
        set<vector<int>> obstacles_set;
        for (vector<int> &obstacle : obstacles)
        {
            obstacles_set.insert(obstacle);
        }

        int x = 0;
        int y = 0;
        int max_squared = x * x + y * y;
        vector<int> robots_vector = {0, 1};

        for (int i = 0; i < commands.size(); i++)
        {
            if (commands[i] == -2)
            {
                vector<int> next_robots_vector = robots_vector;
                next_robots_vector[0] = -1 * robots_vector[1];
                next_robots_vector[1] = robots_vector[0];
                robots_vector = next_robots_vector;
            }
            else if (commands[i] == -1)
            {
                vector<int> next_robots_vector = robots_vector;
                next_robots_vector[0] = robots_vector[1];
                next_robots_vector[1] = -1 * robots_vector[0];
                robots_vector = next_robots_vector;
            }
            else
            {
                int next_x = x;
                int next_y = y;
                for (int j = 0; j < commands[i]; j++)
                {
                    next_x = x + 1 * robots_vector[0];
                    next_y = y + 1 * robots_vector[1];
                    vector<int> next_position = {next_x, next_y};
                    if (obstacles_set.find(next_position) != obstacles_set.end())
                        break;
                    x = next_x;
                    y = next_y;
                    max_squared = max(x * x + y * y, max_squared);
                }
            }
        }

        return max_squared;
    }
};

File: 0874. Walking Robot Simulation/0874.myans.rs
use std::collections::HashSet;

impl Solution {
    pub fn robot_sim(commands: Vec<i32>, obstacles: Vec<Vec<i32>>) -> i32 {
        let obstacles_set: HashSet<(i32, i32)> = obstacles
            .into_iter()
            .map(|v| (v[0], v[1]))
            .collect();

        let dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        let mut dir = 0usize;
        let mut current_step = (0, 0);
        let mut max_square = 0;

        for command in commands {
            if command == -1 {
                dir = (dir + 1) % 4;
                continue;
            }

            if command == -2 {
                dir = (dir + 3) % 4;
                continue;
            }

            let (dx, dy) = dirs[dir];

            for step in 0..command as usize {
                let next_step = (
                    current_step.0 + dx,
                    current_step.1 + dy
                );

                if obstacles_set.contains(&next_step) { break; }

                current_step = next_step;
            }

            let current_square = current_step.0 * current_step.0 + current_step.1 * current_step.1;
            max_square = max_square.max(current_square);
        }

        max_square
    }
}